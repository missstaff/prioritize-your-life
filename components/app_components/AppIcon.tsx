import React from "react";
import { View } from "react-native";
import { ScaledSheet, s } from "react-native-size-matters";
import Ionicons from "@expo/vector-icons/Ionicons";
import { type IconProps } from "@expo/vector-icons/build/createIconSet";
import { type ComponentProps } from "react";

/**
 * Renders an app icon component.
 *
 * @param style - The style object for the component.
 * @param rest - Additional props for the component.
 * @returns The rendered app icon component.
 */
export function AppIcon({
  style,
  ...rest
}: IconProps<ComponentProps<typeof Ionicons>["name"] | any>): JSX.Element {
  return (
    <View style={[styles.container, style]}>
      <Ionicons size={s(28)} {...rest} />
    </View>
  );
}

const styles = ScaledSheet.create({
  container: {
    alignItems: "center",
    backgroundColor: "transparent",
    borderRadius: 100,
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    padding: s(2.5),
  },
});
