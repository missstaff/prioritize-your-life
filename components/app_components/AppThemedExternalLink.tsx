import { Href, Link } from "expo-router";
import { Platform } from "react-native";
import { openBrowserAsync } from "expo-web-browser";
import { type ComponentProps } from "react";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORTHEME } from "@/constants/Colors";

type Props = Omit<ComponentProps<typeof Link>, "href"> & {
  href: Href<string | object>;
};

const AppThemedExternalLink = ({ href, style, ...rest }: Props) => {
  const color = useThemeColor(
    { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
    "background"
  );
  return (
    <Link
      target="_blank"
      {...rest}
      href={href}
      onPress={async (event) => {
        if (Platform.OS !== "web") {
          // Prevent the default behavior of linking to the default browser on native.
          event.preventDefault();
          // Open the link in an in-app browser.
          await openBrowserAsync(href as string);
        }
      }}
      style={[{ color }, style]}
    />
  );
};

export default AppThemedExternalLink;
