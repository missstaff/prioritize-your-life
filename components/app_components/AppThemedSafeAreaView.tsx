import { COLORTHEME } from "@/constants/Colors";
import { useThemeColor } from "@/hooks/useThemeColor";
import { SafeAreaView } from "react-native";


const AppThemedSafeAreaView  = ({ children }: { children: React.ReactNode }) => {
    const backgroundColor = useThemeColor(
        { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
        "background"
      );
    return (
        <SafeAreaView style={{flex: 1, backgroundColor: backgroundColor}}>
            {children}
        </SafeAreaView>
    );
};
export default AppThemedSafeAreaView;