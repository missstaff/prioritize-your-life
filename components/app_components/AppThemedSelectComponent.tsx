
const AppThemedSelectComponent = () => {};

export default AppThemedSelectComponent;