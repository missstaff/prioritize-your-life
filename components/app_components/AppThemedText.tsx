import { Text, type TextProps } from "react-native";
import { ScaledSheet, s } from "react-native-size-matters";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORS, COLORTHEME } from "@/constants/Colors";

export type ThemedTextProps = TextProps & {
  type?: "default" | "title" | "defaultSemiBold" | "subtitle" | "link";
};

const AppThemedText = ({
  style,
  type = "default",
  ...rest
}: ThemedTextProps) => {
  const color = useThemeColor(
    { light: COLORTHEME.light.text, dark: COLORTHEME.dark.text },
    "text"
  );

  return (
    <Text
      style={[
        { color },
        type === "default" ? styles.default : undefined,
        type === "title" ? styles.title : undefined,
        type === "defaultSemiBold" ? styles.defaultSemiBold : undefined,
        type === "subtitle" ? styles.subtitle : undefined,
        type === "link" ? styles.link : undefined,
        style,
      ]}
      {...rest}
    />
  );
};

const styles = ScaledSheet.create({
  default: {
    fontSize: s(16),
    lineHeight: s(24),
  },
  defaultSemiBold: {
    fontSize: s(16),
    fontWeight: "600",
    lineHeight: s(24),
  },
  title: {
    fontSize: s(32),
    fontWeight: "bold",
    lineHeight: s(32),
  },
  subtitle: {
    fontSize: s(20),
    fontWeight: "bold",
  },
  link: {
    color: COLORS.primary,
    fontSize: s(16),
    lineHeight: s(30),
  },
});

export default AppThemedText;
