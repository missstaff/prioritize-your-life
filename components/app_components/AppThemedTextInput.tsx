import React, { useState } from "react";
import {
  StyleProp,
  TextInput,
  TextStyle,
  TouchableOpacity,
  ViewStyle,
  Platform,
} from "react-native";
import { DateTimePickerAndroid } from "@react-native-community/datetimepicker";
import { ScaledSheet, s } from "react-native-size-matters";
import AppIcon from "./AppIcon";
import AppThemedView from "./AppThemedView";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORTHEME } from "@/constants/Colors";
import { TransactionState } from "@/store/transaction/transaction-reducer";

interface AppThemedTextInputProps {
  containerStyle?: StyleProp<ViewStyle>;
  data?: any[] | undefined;
  iconName?: string;
  inputStyle?: StyleProp<TextStyle>;
  keyboardType?: "default" | "numeric";
  placeholder: string;
  secureEntry: boolean;
  value: string;
  checkValue: (value: string) => void;
  setValue: (value: string) => void;
}

const AppThemedTextInput = ({
  containerStyle,
  data,
  inputStyle,
  iconName,
  keyboardType,
  placeholder,
  secureEntry,
  value,
  checkValue,
  setValue,
  ...rest
}: AppThemedTextInputProps) => {
  const textColor = useThemeColor({}, "text");
  let backgroundColor = useThemeColor(
    { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
    "background"
  );

  const [isPasswordVisible, setIsPasswordVisible] = useState(!secureEntry);

  const handleCalendarPress = () => {
    if (Platform.OS === "android") {
      DateTimePickerAndroid.open({
        mode: "date",
        value: new Date(),
        onChange: (event, selectedDate) => {
          if (selectedDate) {
            setValue(formatDate(selectedDate));
          }
        },
      });
    }
  };

  return (
    <AppThemedView style={[styles.inputContainer, containerStyle]}>
      <TextInput
        keyboardType={keyboardType}
        onBlur={(e) => {
          if (value.length > 0) checkValue(value);
        }}
        onChangeText={(text) => {
          setValue(text);
        }}
        placeholder={placeholder}
        placeholderTextColor="#999"
        secureTextEntry={!isPasswordVisible}
        style={[
          styles.input,
          { backgroundColor: backgroundColor, color: textColor },
          inputStyle,
        ]}
        defaultValue={value}
        value={value}
        {...rest}
      />
      {secureEntry && (
        <TouchableOpacity
          onPress={() => setIsPasswordVisible(!isPasswordVisible)}
          testID="passwordVisibilityToggle"
        >
          <AppIcon
            name={!isPasswordVisible ? "eye-off" : "eye"}
            size={s(24)}
            color="#ccc"
          />
        </TouchableOpacity>
      )}

      {iconName && (
        <TouchableOpacity
          disabled={iconName !== "calendar"}
          onPress={() => iconName === "calendar" && handleCalendarPress()}
        >
          <AppIcon name={iconName} size={s(24)} color="#ccc" />
        </TouchableOpacity>
      )}
    </AppThemedView>
  );
};

const styles = ScaledSheet.create({
  inputContainer: {
    alignItems: "center",
    borderColor: "#ccc",
    borderRadius: s(5),
    borderWidth: s(1),
    flexDirection: "row",
    marginBottom: s(10),
    paddingHorizontal: s(10),
    width: "80%",
  },
  input: {
    flex: 1,
    padding: s(10),
  },
});

const formatDate = (date: Date): string => {
  const day = String(date.getDate()).padStart(2, "0");
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const year = date.getFullYear();
  return `${month}/${day}/${year}`;
};

export default AppThemedTextInput;
