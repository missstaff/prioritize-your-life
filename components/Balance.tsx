import { TransactionState } from "@/store/transaction/transaction-reducer";
import AppThemedText from "./app_components/AppThemedText";
import ShowIf from "./ShowIf";
import { ScaledSheet } from "react-native-size-matters";
import { View } from "react-native";

export interface AddTransactionProps {
  balance?: number | undefined;
  data?: any[] | undefined;
  setIsVisible: (isVisible: boolean) => void;
}

const Balance = ({ balance, data, setIsVisible }: AddTransactionProps) => {
  return (
    <View style={styles.container}>
      <ShowIf
        condition={data && data?.length > 0 && balance !== undefined}
        render={
          <AppThemedText type="default">
            {`$${balance?.toFixed(2)}`}
          </AppThemedText>
        }
        renderElse={<View></View>}
      />
      <ShowIf
        condition={data && data?.length > 0}
        render={
          <AppThemedText type="link" onPress={() => [setIsVisible(true)]}>
            Add {data && data[0]?.goal ? "Goal" : "Transaction"}
          </AppThemedText>
        }
        renderElse={<View></View>}
      />
    </View>
  );
};

const styles = ScaledSheet.create({
  container: {
    alignItems: "center",
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-around",
    width: "100%",
  },
});

export default Balance;
