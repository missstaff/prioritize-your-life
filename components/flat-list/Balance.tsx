import { TransactionState } from "@/store/transaction/transaction-reducer";
import AppThemedText from "../app_components/AppThemedText";
import ShowIf from "../ShowIf";
import AppThemedView from "../app_components/AppThemedView";

export interface AddTransactionProps {
  balance: number | undefined;
  data: TransactionState[] | undefined;
  setIsVisible: (isVisible: boolean) => void;
}

/**
 * Represents the balance component.
 * @param {AddTransactionProps} props - The props for the balance component.
 * @param {number | undefined} props.balance - The balance to be displayed.
 * @param {TransactionState[] | undefined} props.data - The transaction data.
 * @param {function} props.setIsVisible - The function to set the visibility of the component.
 * @returns {JSX.Element} The rendered balance component.
 */
export default function Balance({
  balance,
  data,
  setIsVisible,
}: AddTransactionProps) {
  return (
    <>
      <AppThemedText>
        {!data?.length ? "$0.00" : "$" + balance?.toFixed(2)}
      </AppThemedText>
      <ShowIf
        condition={data && data?.length > 0}
        render={
          <AppThemedText type="link" onPress={() => [setIsVisible(true)]}>
            Add Transaction
          </AppThemedText>
        }
        renderElse={
         <AppThemedView></AppThemedView>
        }
      />
    </>
  );
}
