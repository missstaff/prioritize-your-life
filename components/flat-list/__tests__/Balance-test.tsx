import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import Balance, { AddTransactionProps } from "../Balance";
import { AppThemedText } from "../../app_components/AppThemedText";

jest.mock("../../app_components/AppThemedText", () => ({
  AppThemedText: jest.fn(({ children, onPress }) => (
    <div onClick={onPress}>{children}</div>
  )),
}));

describe("Balance Component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly with no data (shows $0.00)", () => {
    const mockSetIsVisible = jest.fn();

    const props: AddTransactionProps = {
      balance: 0,
      data: undefined,
      setIsVisible: mockSetIsVisible,
    };

    const { toJSON } = render(<Balance {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with data", () => {
    const mockSetIsVisible = jest.fn();

    const props: AddTransactionProps = {
      balance: 100,
      data: [
        {
          id: "1",
          amount: "100",
          date: new Date().toUTCString(),
          description: "Test",
          balance: 100,
        },
      ],
      setIsVisible: mockSetIsVisible,
    };

    const { toJSON } = render(<Balance {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders the balance correctly", () => {
    const mockSetIsVisible = jest.fn();

    const props: AddTransactionProps = {
      balance: 100,
      data: [
        {
          id: "1",
          amount: "100",
          date: new Date().toUTCString(),
          description: "Test",
          balance: 100,
        },
      ],
      setIsVisible: mockSetIsVisible,
    };

    render(<Balance {...props} />);

    // Assert that AppThemedText was called with the correct balance text
    expect(AppThemedText).toHaveBeenCalledWith(
      expect.objectContaining({
        children: "$100.00",
      }),
      {}
    );
  });

  //   it("renders Add Transaction link and triggers setIsVisible", () => {
  //     const mockSetIsVisible = jest.fn();

  //     const props: AddTransactionProps = {
  //       balance: 100,
  //       data: [
  //         {
  //           id: "1",
  //           amount: "100",
  //           date: new Date().toUTCString(),
  //           description: "Test",
  //           balance: 100,
  //         },
  //       ],
  //       setIsVisible: mockSetIsVisible,
  //     };

  //     const { getByText } = render(<Balance {...props} />);

  //     // Check if "Add Transaction" text is rendered
  //     const addTransactionLink = getByText("Add Transaction");

  //     // Simulate a press on the "Add Transaction" link
  //     fireEvent.press(addTransactionLink);

  //     // Ensure setIsVisible was called
  //     expect(mockSetIsVisible).toHaveBeenCalledWith(true);
  //   });
});
