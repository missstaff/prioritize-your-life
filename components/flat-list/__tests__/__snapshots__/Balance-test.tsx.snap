// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balance Component renders correctly with data 1`] = `
[
  <div>
    $100.00
  </div>,
  <div
    onClick={[Function]}
  >
    Add Transaction
  </div>,
]
`;

exports[`Balance Component renders correctly with no data (shows $0.00) 1`] = `
[
  <div>
    $0.00
  </div>,
  <div
    onClick={[Function]}
  >
    Add Transaction
  </div>,
]
`;
